### PR TITLE
SQLite state store

### PR DESCRIPTION
**Pull request by zecakeh**
Supercedes #1458.
Based on #1798.

Signed-off-by: Kévin Commaille <zecakeh@tedomum.fr>

### CODE REVIEW

**Comment by zecakeh** [comment]
Rebased on main.

File: crates/matrix-sdk-sqlite/migrations/state_store/001_init.sql [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    PRIMARY KEY ("room_id", "name")
+);
+
+CREATE TABLE "media" (
+    "uri" BLOB NOT NULL,
+    "format" BLOB NOT NULL,
+    "data" BLOB NOT NULL,
+
+    PRIMARY KEY ("uri", "format")
+);
```
Comment (jplatte) [inline crates/matrix-sdk-sqlite/migrations/state_store/001_init.sql:98]: Missing final newline

File: crates/matrix-sdk-sqlite/src/state_store.rs [hunk 2]
```diff
@@ -0,0 +1,1426 @@
...
+    fn remove_room_receipts(&self, room_id: &[u8]) -> rusqlite::Result<()>;
+
+    fn set_display_name(&self, room_id: &[u8], name: &[u8], data: &[u8]) -> rusqlite::Result<()>;
+    fn remove_display_name(&self, room_id: &[u8], name: &[u8]) -> rusqlite::Result<()>;
+    fn remove_room_display_names(&self, room_id: &[u8]) -> rusqlite::Result<()>;
+}
+
+impl SqliteConnectionStateStoreExt for rusqlite::Connection {
+    fn set_kv_blob(&self, key: &[u8], value: &[u8]) -> rusqlite::Result<()> {
+        self.execute("INSERT OR REPLACE INTO kv_blob VALUES (?1, ?2)", (key, value))?;
```
Comment (jplatte) [inline crates/matrix-sdk-sqlite/src/state_store.rs:315]: No need to number the parameters anymore with `INSERT OR REPLACE`.

```suggestion
        self.execute("INSERT OR REPLACE INTO kv_blob VALUES (?, ?)", (key, value))?;
```

File: crates/matrix-sdk-sqlite/src/state_store.rs [hunk 3]
```diff
@@ -0,0 +1,1426 @@
...
+        Ok(())
+    }
+
+    fn set_global_account_data(&self, event_type: &[u8], data: &[u8]) -> rusqlite::Result<()> {
+        self.prepare_cached(
+            "INSERT OR REPLACE INTO global_account_data (event_type, data)
+             VALUES (?1, ?2)",
```
Comment (jplatte) [inline crates/matrix-sdk-sqlite/src/state_store.rs:322]: Same here, and in more functions below.

File: crates/matrix-sdk-sqlite/src/state_store.rs [hunk 4]
```diff
@@ -0,0 +1,1426 @@
...
+                |row| Ok((row.get(0)?, row.get(1)?)),
+            )
+            .await
+            .optional()?)
+    }
+
+    async fn get_state_events(&self, room_id: Key, event_type: Key) -> Result<Vec<Vec<u8>>> {
+        Ok(self
+            .prepare(
+                "SELECT data FROM state_event WHERE room_id = ? AND event_type = ? AND stripped = 0",
```
Comment (jplatte) [inline crates/matrix-sdk-sqlite/src/state_store.rs:588]:
```suggestion
                "SELECT data FROM state_event WHERE room_id = ? AND event_type = ? AND stripped = FALSE",
```

just for readability

File: crates/matrix-sdk-sqlite/src/state_store.rs [hunk 5]
```diff
@@ -0,0 +1,1426 @@
...
+            self.prepare("SELECT data FROM member WHERE room_id = ?", |mut stmt| {
+                stmt.query((room_id,))?.mapped(|row| row.get(0)).collect()
+            })
+            .await?
+        } else {
+            let sql_params = vec!["?"; memberships.len()].join(", ");
+            let sql = format!(
+                "SELECT data FROM member WHERE room_id = ? AND membership IN ({sql_params})"
+            );
+            memberships.insert(0, room_id);
```
Comment (jplatte) [inline crates/matrix-sdk-sqlite/src/state_store.rs:616]: This looks weird to me. How about

```rust
let params = iter::once(room_id).chain(memberships);
```

instead?

File: crates/matrix-sdk-sqlite/src/state_store.rs [hunk 6]
```diff
@@ -0,0 +1,1426 @@
...
+
+    async fn save_changes(&self, changes: &StateChanges) -> Result<()> {
+        let changes = changes.to_owned();
+        let this = self.clone();
+        self.acquire()
+            .await?
+            .with_transaction(move |txn| {
+                let StateChanges {
+                    sync_token,
+                    session: _,
```
Comment (jplatte) [inline crates/matrix-sdk-sqlite/src/state_store.rs:797]: Do the other stores also ignore this one?

File: crates/matrix-sdk-sqlite/src/state_store.rs [hunk 7]
```diff
@@ -0,0 +1,1426 @@
...
+                    profiles,
+                    state,
+                    room_account_data,
+                    room_infos,
+                    receipts,
+                    redactions,
+                    stripped_state,
+                    stripped_room_infos,
+                    ambiguity_maps,
+                    notifications: _,
```
Comment (jplatte) [inline crates/matrix-sdk-sqlite/src/state_store.rs:809]: Same Q.

File: crates/matrix-sdk-sqlite/src/state_store.rs (hunk 6 above)
Comment (zecakeh) [reply crates/matrix-sdk-sqlite/src/state_store.rs:797]: Yes, I'm not event sure why it's in StateChanges.

File: crates/matrix-sdk-sqlite/src/state_store.rs (hunk 7 above)
Comment (zecakeh) [reply crates/matrix-sdk-sqlite/src/state_store.rs:809]: Yes, it's only used by the notifications handler

File: crates/matrix-sdk-sqlite/src/utils.rs [hunk 8]
```diff
@@ -48,7 +55,11 @@ pub(crate) trait SqliteObjectExt {
 
     async fn execute_batch(&self, sql: &'static str) -> rusqlite::Result<()>;
 
-    async fn prepare<T, F>(&self, sql: &'static str, f: F) -> rusqlite::Result<T>
+    async fn prepare<T, F>(
+        &self,
+        sql: impl AsRef<str> + Send + 'static,
```
[the remark this answers is not in the document]
Comment (jplatte) [reply crates/matrix-sdk-sqlite/src/utils.rs:60]: I wouldn't mind.

**Review by jplatte — APPROVED** [review]

**Pull request merged**